File: iOS/Pulsar/Sources/Pulsar/Pulsar.swift
import CoreHaptics
import UIKit

@objc public class Pulsar: NSObject {
  private var engine = HapticEngineWrapper()
  private var audioSimulator = AudioSimulator()
  private var presets: PresetsWrapper?
  private var realtimeComposer: RealtimeComposerImpl!
  
  @objc public override init() {
    realtimeComposer = RealtimeComposerImpl(engine: self.engine)
  }
  
  @objc public func Presets() -> PresetsWrapper {
    if (presets == nil) {
      presets = PresetsWrapper(haptics: self)
    }
    return presets!
  }
  
  @objc public func preloadPresets(presetNames: Array<String>) {
    let presets = self.Presets()
    for (presetName) in presetNames {
      presets.preloadPresetByName(presetName)
    }
  }
  
  @objc public func enableSound(state: Bool) {
    audioSimulator.enableSound(state)
  }
  
  @objc public func enableCache(state: Bool) {
    self.Presets().enableCache(state: state)
  }
  
  @objc public func clearChace() {
    self.Presets().resetCache()
  }
  
  @objc public func PatternComposer() -> PatternComposerImpl {
    return PatternComposerImpl(engine: engine, audioSimulator: audioSimulator)
  }
  
  @objc public func RealtimeComposer() -> RealtimeComposerImpl {
    return realtimeComposer
  }
  
}

File: Android/Pulsar/src/main/java/com/swmansion/pulsar/Pulsar.kt
package com.swmansion.pulsar

import android.content.Context
import com.swmansion.pulsar.audio.AudioSimulator
import com.swmansion.pulsar.composers.PatternComposer
import com.swmansion.pulsar.composers.RealtimeComposer
import com.swmansion.pulsar.haptics.HapticEngineWrapper
import com.swmansion.pulsar.presets.PresetsWrapper
import com.swmansion.pulsar.types.CompatibilityMode
import com.swmansion.pulsar.types.RealtimeComposerStrategy

class Pulsar(private var context: Context) {
    private val engine = HapticEngineWrapper(context)
    private val audioSimulator = AudioSimulator(engine.getRealCompatibilityMode())
    private var presets: PresetsWrapper? = null
    private var realtimeComposer: RealtimeComposer? = null

    fun getPresets(): PresetsWrapper {
        if (presets == null) {
            presets = PresetsWrapper(this, context, engine)
        }
        return presets!!
    }

    fun getPatternComposer(): PatternComposer {
        return PatternComposer(engine, audioSimulator)
    }

    fun getRealtimeComposer(strategy: RealtimeComposerStrategy = RealtimeComposerStrategy.ENVELOPE): RealtimeComposer {
        if (realtimeComposer == null) {
            realtimeComposer = RealtimeComposer(engine, strategy)
        }
        return realtimeComposer!!
    }

    fun hapticSupport(): CompatibilityMode {
        return engine.getRealCompatibilityMode()
    }

    fun forceHapticsSupportLevel(mode: CompatibilityMode) {
        engine.simulateCompatibilityMode(mode)
        audioSimulator.setCompatibilityMode(mode)
    }

    fun preloadPresets(presetNames: List<String>) {
        this.getPresets().preloadPresetByNames(presetNames)
    }

    fun enableHaptics(state: Boolean) {
        engine.enableHaptics(state)
    }

    fun enableSound(state: Boolean) {
        audioSimulator.enableSound(state)
    }

    fun enableCache(state: Boolean) {
        this.getPresets().enableCache(state)
    }

    fun clearCache() {
        this.getPresets().resetCache()
    }

    fun stopHaptics() {
        engine.stop()
    }
}